---
title: "An Introduction to RNA-sequencing"
subtitle: "Differential Expression and Pathway Analysis"
author: | 
  | W. Evan Johnson, Ph.D.
  | Professor, Division of Infectious Disease
  | Director, Center for Data Science
  | Co-Director, Center for Biomedical Informatics and Health AI
  | Rutgers University -- New Jersey Medical School
date: "`r Sys.Date()`"
header-includes:
  - \usepackage{amsmath}
  - \usepackage{xcolor}
  - \setbeamercolor{frametitle}{fg=black}
  - \usepackage{graphicx}
  - \usebackgroundtemplate{\includegraphics[width=\paperwidth]{regfigs/RH_template_Page_2.png}}
  - \addtobeamertemplate{frametitle}{\vspace*{.25in}}{\vspace*{.25in}}
  - \setbeamerfont{frametitle}{size=\huge}
  - \usepackage{tikz}
output: 
  beamer_presentation
classoption: aspectratio=169 
editor_options: 
  chunk_output_type: console
tables: true
---

## Load Packages for Today
\large
We will be using the following packages for today's lecture:
We will be using the following packages for our RNA-seq lecture:
```{r, warning=FALSE, message=FALSE}
library(tidyverse) ## tools for data wranging
library(SummarizedExperiment) ## managing counts data
library(edgeR) ## differential expression
library(DESeq2) ## differential expression
library(ComplexHeatmap) ## Heatmap visualization
library(TBSignatureProfiler) ## TB signature analysis
library(umap) ## dimenstion reduction and plotting data
```


## Installing and using the SCTK
\Large
```{r, eval=F}
install.packages("devtools")
devtools::install_github("wevanjohnson/singleCellTK")
library(singleCellTK)
singleCellTK()

### Example: open example_data/ 
### features_combined.txt and meta_data.txt
```


## Background

\Large
- RNA-seq measures gene expression as count data.  
- Count data are overdispersed: variance >> mean.  
- Poisson regression assumes equal mean and variance â€” often violated.  
- Use the **Negative Binomial (NB)** model to allow for overdispersion.

---

## Negative Binomial Model

\Large
Let $K_{ij}$ be the count for gene $i$ in sample $j$.

\[
K_{ij} \sim \text{NB}(\mu_{ij}, \alpha_i)
\]

- $\mu_{ij}$: expected count  
- $\alpha_i$: dispersion parameter for gene $i$

Variance:

\[
\text{Var}(K_{ij}) = \mu_{ij} + \alpha_i \mu_{ij}^2
\]

---

## Link Function and Design Matrix

\Large
We model the expected count using a log link:

\[
\log(\mu_{ij}) = \log(s_j) + \mathbf{x}_j^\top \boldsymbol{\beta}_i
\]

Where:
- $s_j$: size factor (normalizes for sequencing depth)  
- $\mathbf{x}_j$: covariates (e.g., condition, batch)  
- $\boldsymbol{\beta}_i$: regression coefficients for gene $i$

---

## Estimating Dispersion
\Large

- Dispersion $\alpha_i$ is typically gene-specific.  
- Estimated via:
  - Empirical Bayes shrinkage (DESeq2)  
  - Tagwise/Moderated dispersion (edgeR)  

Goal: Stabilize estimates for low-count genes across the genome.

---

## Hypothesis Testing

\Large
For each gene, test:

\[
H_0: \beta_{i1} = 0 \quad \text{(no differential expression)}
\]

Common methods:
- Wald Test (DESeq2)  
- Likelihood Ratio Test (edgeR, DESeq2)

Adjust for multiple testing:

\[
\text{FDR control using Benjamini-Hochberg}
\]

---

## Summary of RNA-seq Analysis

\Large
- Negative Binomial models handle overdispersion in RNA-seq.  
- Log-linear link connects expression to experimental design.  
- Shrinkage improves dispersion estimation.  
- Differential expression is tested gene-wise with multiple testing correction.

## Make a SummarizedExperiment
\Large
Using an example dataset from: [Verma, et al., 2018](https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-018-3127-4)
```{r}
## read in data
counts <- read.table(
  "example_data/features_combined.txt", 
  sep="\t", header=T, row.names=1)
meta_data <- read.table(
  "example_data/meta_data.txt",
  sep="\t", header=T, row.names=1)
group <- meta_data$Disease
```

## Make a SummarizedExperiment
```{r}
## Make SummarizedExperiment
se_hivtb <- SummarizedExperiment(assays=list(counts=counts),
                     colData = meta_data)

## Make log counts, counts per million (cpm), logcpm
se_hivtb <- mkAssay(se_hivtb, log = TRUE, 
                     counts_to_CPM = TRUE)
assays(se_hivtb)
```

## EdgeR Example
Implements statistical methods for DE analysis based on the negative binomial model:

```{r}
#Gene Filtering
counts<-counts[which(rowSums(counts)>100),] 
#Computes library size
dge <- DGEList(counts=counts, group=group) 
#TMM normalization
dge <- calcNormFactors(dge) 
# Design matrix
design<-model.matrix(~Disease, data=meta_data)
#Estimates common, trended and tagwise dispersion
dge<-estimateDisp(counts,design) 
```

## EdgeR Example
\Large
In negative binomial models, each gene is given a dispersion parameter.  Dispersions control the variances of the gene counts and underestimation will lead to false discovery and overestimation may lead to a lower rate of true discovery.

## EdgeR Example
\large
```{r}
# Neg Bin GLM with the dispersion estimates
fit<-glmFit(counts,design, 
            dispersion=dge$tagwise.dispersion) 
# Performs likelihood ratio test
# Compares full versus reduced model
lrt<-glmLRT(fit, coef=2) 
```


## EdgeR Example

\small
```{r}
# Prints the top results
topTags(lrt)
```

## EdgeR Example
\Large
```{r}
## Use the a quasi-likelihood F-test 
fit<-glmQLFit(counts, design, 
        dispersion=dge$tagwise.dispersion) 
## use for small datasets 
qlf<-glmQLFTest(fit, coef=2)
```

## EdgeR Example
\small
```{r}
# Prints the top results
topTags(qlf)
```

## EdgeR Example
\Large
```{r}
#For visualization, heatmaps/PCA
Logcpm<-cpm(counts,log=TRUE)
```

## DESeq2 Example
```{r, include=F}
meta_data$Disease = as.factor(meta_data$Disease)
```
\Large
```{r}
#colData is a data frame of demographic/phenotypic data
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData=meta_data, 
                              design=~Disease)
#Gene Filtering
dds<-dds[rowSums(counts(dds))>100,] 
```

## DESeq2 Example
\scriptsize
```{r}
#Performs estimation of size factors,
#dispersion, and negative binomial GLM fitting
dds<-DESeq(dds) 
```

## DESeq2 Example
\scriptsize
```{r}
res <- results(dds)[order(results(dds)[,6]),]
res[1:25,]
``` 

## Heatmap of DEGs
```{r, eval=F}
# Make a Heatmap of DEGs
mat = as.matrix(assay(se_hivtb,"counts")
                )[order(results(dds)[,6])[1:100],] 
                # Using first 100 genes to simplify
mat = t(scale(t(mat))) ## scale gene expression
df=data.frame(Disease=colData(se_hivtb)$Disease) 
ha = HeatmapAnnotation(df = df, 
                       col = list(Disease=c(
                         "tb_hiv"="Red",
                         "hiv_only"="Blue",
                         "tb_hiv_art"="Green")))
Heatmap(mat,show_row_names=F, show_column_names = F, 
        top_annotation = ha)
```

## Heatmap of DEGs

```{r, echo=F, out.height='70%',fig.align='center'}
# Make a Heatmap of DEGs
mat = as.matrix(assay(se_hivtb,"log_counts_cpm")
                )[order(results(dds)[,6])[1:100],] 
                # Using first 1000 genes to simplify
mat = t(scale(t(mat)))
df=data.frame(Disease=colData(se_hivtb)$Disease) 
ha = HeatmapAnnotation(df = df, 
                       col = list(Disease=c(
                         "tb_hiv"="Red",
                         "hiv_only"="Blue",
                         "tb_hiv_art"="Green")))
Heatmap(mat,show_row_names=F,
        show_column_names = F, 
        top_annotation = ha)
```


## Limma Example
\Large
* Most similar to microarray data flow
* Reads counts are converted to log2 counts per million (logCPM) and the mean-variance relationship is modeled with precision weights (voom transform)

## Limma Example
```{r}
#From edgeR, Computes library size
dge <- DGEList(counts=counts, group=group) 
#Gene Filtering
counts<-counts[which(rowSums(cpm(counts))>1),] 
dge <- calcNormFactors(dge) #TMM normalization
```

## Limma Example
\Large
```{r}
design<-model.matrix(~group)
#voom transform to calculate weights to 
#eliminate mean-variance relationship
v<-voom(dge, design)
#use usual limma pipelines
fit<-lmFit(v,design)
fit<-eBayes(fit)
```

## Limma Example
\scriptsize
```{r}
topTable(fit, coef=ncol(design))
```


## Pathway analysis
\Large
After finding DEGs, look for correlated genes/networks and enriched pathway sets in the gene set using: 

* Weighted gene coexpression network analysis (WGCNA)
* GSEA, GSVA, EnrichR, many more!!
* Qiagen Ingenuity Pathway Analysis (IPA)

## Pathway analysis
\center
![](pathfigs/pathway1.png){height="70%"}

## Pathway analysis
\center
![](pathfigs/pathway2.png){height=70%}

## TBSignatureProfiler Analysis
\Large
The TBSignatureProfiler was developed in the Johnson Lab in 2021 to profile new and existing TB gene expression signatures:

[https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-020-05598-z](https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-020-05598-z)

## TBSignatureProfiler Analysis
\Large
```{r, message=FALSE, results='hide'}
se_hivtb_2 <- se_hivtb[,
      colData(se_hivtb)$Disease != "tb_hiv_art"]
TBsigs <- TBsignatures[-12] 
ssgsea_res <- runTBsigProfiler(se_hivtb_2, 
                  useAssay = "log_counts_cpm",
                  signatures = TBsigs,
                  algorithm = "ssGSEA",
                  combineSigAndAlgorithm = TRUE,
                  parallel.sz = 1)
```

## Signature Heatmap:
\Large
```{r, eval=F}
# Colors for gradient
signatureHeatmap(ssgsea_res, 
        name = "Heatmap of Signatures (ssGSEA)", 
        signatureColNames = names(TBsigs),
        annotationColNames = c("Disease"),
        scale = TRUE,
        split_heatmap = "none",
        showColumnNames = FALSE)
```

## Signature Heatmap:
```{r, echo=F, out.height="80%", fig.align='center'}
# Colors for gradient
signatureHeatmap(ssgsea_res, 
        name = "Heatmap of Signatures (ssGSEA)", 
        signatureColNames = names(TBsigs),
        annotationColNames = c("Disease"),
        scale = TRUE,
        split_heatmap = "none",
        showColumnNames = FALSE)
```

## Signature Boxplots
```{r, warning=F, out.height="50%", fig.align='center'}
signatureBoxplot(ssgsea_res, name="ssGSEA", 
          signatureColNames = names(TBsigs)[c(62,77)],
          annotationColName = c("Disease"))
```

## Session info
\tiny
```{r session info}
sessionInfo()
```
