---
title: "An Introduction to RNA-sequencing"
subtitle: "Differential Expression Analysis"
author: | 
  | W. Evan Johnson, Ph.D.
  | Professor, Division of Infectious Disease
  | Director, Center for Data Science
  | Co-Director, Center for Biomedical Informatics and Health AI
  | Rutgers University -- New Jersey Medical School
date: "`r Sys.Date()`"
header-includes:
  - \usepackage{amsmath}
  - \usepackage{xcolor}
  - \setbeamercolor{frametitle}{fg=black}
  - \usepackage{graphicx}
  - \usebackgroundtemplate{\includegraphics[width=\paperwidth]{degfigs/RH_template_Page_2.png}}
  - \addtobeamertemplate{frametitle}{\vspace*{.25in}}{\vspace*{.25in}}
  - \setbeamerfont{frametitle}{size=\huge}
  - \usepackage{tikz}
output: 
  beamer_presentation
classoption: aspectratio=169 
editor_options: 
  chunk_output_type: console
tables: true
---


## Installing R Packages: 
\Large
Install the following tools: `Rsubread`, `Rsamtools`,  and `SummarizedExperiment`. We will also need help from the `tidyverse.`

```{r eval=F}
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")
BiocManager::install(c("Rsubread","Rsamtools",
             "tidyverse","SummarizedExperiment"))
```


## Load Packages for Today
\large
We will be using the following packages for today's lecture:
We will be using the following packages for our RNA-seq lecture:
```{r, warning=FALSE, message=FALSE}
library(tidyverse) ## tools for data wranging
library(Rsubread) ## alignment and feature counts
library(Rsamtools) ## managing .sam and .bam files
library(SummarizedExperiment) ## managing counts data
library(edgeR) ## differential expression
library(DESeq2) ## differential expression
library(ComplexHeatmap) ## Heatmap visualization
library(TBSignatureProfiler) ## TB signature analysis
library(umap) ## dimenstion reduction and plotting data
```


## Using Rsubread to do Alignment

\Large 
The following userguide will be helpful for you: 

http://bioinf.wehi.edu.au/subread-package/SubreadUsersGuide.pdf


## Indexing your genome

\Large 
Note that you will rarely do this for human alignment. You will usually download an existing index given to you by others who have already done this work. You will do this often if you are aligning microbial reads, e.g. MTB or some other organism for which others have not already made your index for you.

\normalsize
```{r, eval=F}
buildindex(basename=".../lecture_1/example_data/genome/ucsc.hg19.chr1_120-150M",
 reference="../lecture_1/example_data/genome/ucsc.hg19.chr1_120-150M.fasta.gz")
```

\Large
Took me ~0.2 minutes!

## Aligning your reads:  
\Large
Note that this outputs results in a .bam file and not a .sam file

\normalsize
```{r, eval=F, include=F}
align(index="example_data/genome/ucsc.hg19.chr1_120-150M",readfile1="../lecture1/example_data/reads/R01_10_short500K.fq.gz",output_file="../lecture1/example_data/alignments/R01_10_short.bam", nthreads=4)
```

```{r eval=F, include=F}
align(index="rna_seq/genome/ucsc.hg19.chr1_120-150M",readfile1="rna_seq/reads/R01_10_short500K.fq.gz",output_file="rna_seq/alignments/R01_10_short.bam", nthreads=4) 
```

```{r, eval=F}
align(index="rna_seq/genome/ucsc.hg19.chr1_120-150M",
      readfile1="rna_seq/reads/R01_10_short500K.fq.gz",
      output_file="rna_seq/alignments/R01_10_short.bam", 
      nthreads=4) 
```

My old laptop was an Apple M2, with 8 cores (used 4 cores), 24GB RAM: 

  - Took 15.7 minutes to align ~60M reads to the 30M bases
  - Took 0.7 minutes to align ~6.5M reads to the 30M bases
  - Took 0.3 minutes to align ~500K reads to the 30M bases

## Algning the reads using STAR

\Large
This afternoon we will use the STAR aligner on Amarel to align a set of RNA-seq fastq files!

## Aligned Sequencing Data Formats (SAM and BAM)
Note that Rsubread outputs a .bam file (bam = binary alignment map) and not a .sam file (sam = sequence alignment map). Here is some information about a .sam file: https://en.wikipedia.org/wiki/SAM_(file_format)

![](degfigs/sambam.png)

## Aligned Sequencing Data Formats (SAM and BAM)
\Large
To convert .sam to .bam or vice versa, a package called Rsamtools. Using Rsamtools, you can convert bam to sam as follows:

```{r, eval=F}
asSam("rna_seq/alignments/R01_10_short.bam", 
      overwrite=T) 

# To convert to bam:
#asBam("rna_seq/alignments/R01_10_short.bam") 
```


## Feature counts
\Large
Now we can count reads hitting genes. Approaches/software: 

* HT-Seq
* STAR
* Cufflinks
* RPKM FPKM or CPM
* RSEM
* edgeR
* findOverlaps (GenomicRanges)
* featureCounts (Rsubread)

## Feature counts
\center
![](degfigs/featurecounts.png){height=60%}

## Feature counts

```{r, eval=FALSE}
fCountsList = featureCounts(
  "rna_seq/alignments/R01_10_short.bam",
  annot.ext="rna_seq_files/genome/genes.chr1_120-150M.gtf",
  isGTFAnnotationFile=TRUE)

featureCounts = cbind(fCountsList$annotation[,1], 
                      fCountsList$counts)

write.table(featureCounts, 
    "rna_seq/alignments/R01_10_short.features.txt", 
    sep="\t", col.names=FALSE, row.names=FALSE, quote=FALSE)
```

## Data Structures
\Large
A data structure is a particular way of organizing data in a computer so that it can be used effectively. The idea is to reduce the space and time complexities of different tasks.

## Data Structures
\Large
Data structures in R programming are tools for holding multiple values, variables, and sometimes functions

**Please think very carefully about the way you manage and store your data!** This can make your life much easier and make your code and data cleaner and more portable!

## Data Frames
\Large
A large proportion of data analysis challenges start with data stored in a data frame. For example, we stored the data for our motivating example in a data frame. You can access this dataset by loading `TBNanostring.rds` object in R:

```{r}
TBnanostring <- readRDS("example_data/TBnanostring.rds")
```

## Data Frames
\Large
In RStudio we can view the data with the `View` function:

```{r, eval=F}
View(TBnanostring)
```

## Data Frames
\Large
Or in RMarkdown you can use the `datatable` function from the `DT` package:

```{r, eval=F}
datatable(TBnanostring)
```

You will notice that the TB status is found in the first column of the data frame, followed by the genes in the subsequent columns. The rows represent each individual patient. 

## Advanced Data Structures

There are advanced R data structures that can facilitate object orientated programming. One useful example is the __SummarizedExperiment__ object. 

\center
![](degfigs/summarizedexperiment.png){height=60%}

## Make a SummarizedExperiment
\Large
Using an example dataset from: [Verma, et al., 2018](https://bmcinfectdis.biomedcentral.com/articles/10.1186/s12879-018-3127-4)
```{r}
## read in data
counts <- read.table(
  "example_data/features_combined.txt", 
  sep="\t", header=T, row.names=1)
meta_data <- read.table(
  "example_data/meta_data.txt",
  sep="\t", header=T, row.names=1)
group <- meta_data$Disease
```

## Make a SummarizedExperiment
```{r}
## Make SummarizedExperiment
se_hivtb <- SummarizedExperiment(assays=list(counts=counts),
                     colData = meta_data)

## Make log counts, counts per million (cpm), logcpm
se_hivtb <- mkAssay(se_hivtb, log = TRUE, 
                     counts_to_CPM = TRUE)
assays(se_hivtb)
```

## EdgeR Example
Implements statistical methods for DE analysis based on the negative binomial model:

```{r}
#Gene Filtering
counts<-counts[which(rowSums(counts)>100),] 
#Computes library size
dge <- DGEList(counts=counts, group=group) 
#TMM normalization
dge <- calcNormFactors(dge) 
# Design matrix
design<-model.matrix(~Disease, data=meta_data)
#Estimates common, trended and tagwise dispersion
dge<-estimateDisp(counts,design) 
```

## EdgeR Example
\Large
In negative binomial models, each gene is given a dispersion parameter.  Dispersions control the variances of the gene counts and underestimation will lead to false discovery and overestimation may lead to a lower rate of true discovery.

## EdgeR Example
\large
```{r}
# Neg Bin GLM with the dispersion estimates
fit<-glmFit(counts,design, 
            dispersion=dge$tagwise.dispersion) 
# Performs likelihood ratio test
# Compares full versus reduced model
lrt<-glmLRT(fit, coef=2) 
```


## EdgeR Example

\small
```{r}
# Prints the top results
topTags(lrt)
```

## EdgeR Example
\Large
```{r}
## Use the a quasi-likelihood F-test 
fit<-glmQLFit(counts, design, 
        dispersion=dge$tagwise.dispersion) 
## use for small datasets 
qlf<-glmQLFTest(fit, coef=2)
```

## EdgeR Example
\small
```{r}
# Prints the top results
topTags(qlf)
```

## EdgeR Example
\Large
```{r}
#For visualization, heatmaps/PCA
Logcpm<-cpm(counts,log=TRUE)
```

## DESeq2 Example
```{r, include=F}
meta_data$Disease = as.factor(meta_data$Disease)
```

```{r}
#colData is a data frame of demographic/phenotypic data
dds <- DESeqDataSetFromMatrix(countData = counts, 
                              colData=meta_data, 
                              design=~Disease)
#Gene Filtering
dds<-dds[rowSums(counts(dds))>100,] 
```

## DESeq2 Example
\scriptsize
```{r}
#Performs estimation of size factors,
#dispersion, and negative binomial GLM fitting
dds<-DESeq(dds) 
```

## DESeq2 Example
\scriptsize
```{r}
res <- results(dds)[order(results(dds)[,6]),]
res[1:25,]
``` 

## Heatmap of DEGs
```{r, eval=F}
# Make a Heatmap of DEGs
mat = as.matrix(assay(se_hivtb,"counts")
                )[order(results(dds)[,6])[1:100],] 
                # Using first 100 genes to simplify
mat = t(scale(t(mat))) ## scale gene expression
df=data.frame(Disease=colData(se_hivtb)$Disease) 
ha = HeatmapAnnotation(df = df, 
                       col = list(Disease=c(
                         "tb_hiv"="Red",
                         "hiv_only"="Blue",
                         "tb_hiv_art"="Green")))
Heatmap(mat,show_row_names=F, show_column_names = F, 
        top_annotation = ha)
```

## Heatmap of DEGs

```{r, echo=F, out.height='70%',fig.align='center'}
# Make a Heatmap of DEGs
mat = as.matrix(assay(se_hivtb,"log_counts_cpm")
                )[order(results(dds)[,6])[1:100],] 
                # Using first 1000 genes to simplify
mat = t(scale(t(mat)))
df=data.frame(Disease=colData(se_hivtb)$Disease) 
ha = HeatmapAnnotation(df = df, 
                       col = list(Disease=c(
                         "tb_hiv"="Red",
                         "hiv_only"="Blue",
                         "tb_hiv_art"="Green")))
Heatmap(mat,show_row_names=F,
        show_column_names = F, 
        top_annotation = ha)
```


## Limma Example
\Large
* Most similar to microarray data flow
* Reads counts are converted to log2 counts per million (logCPM) and the mean-variance relationship is modeled with precision weights (voom transform)

## Limma Example
```{r}
#From edgeR, Computes library size
dge <- DGEList(counts=counts, group=group) 
#Gene Filtering
counts<-counts[which(rowSums(cpm(counts))>1),] 
dge <- calcNormFactors(dge) #TMM normalization
```

## Limma Example
\Large
```{r}
design<-model.matrix(~group)
#voom transform to calculate weights to 
#eliminate mean-variance relationship
v<-voom(dge, design)
#use usual limma pipelines
fit<-lmFit(v,design)
fit<-eBayes(fit)
```

## Limma Example
\scriptsize
```{r}
topTable(fit, coef=ncol(design))
```

## Session info
\tiny
```{r session info}
sessionInfo()
```
